frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -881,20 +885,22 @@ pub mod pallet {
 
 			// ensure solution claims is better.
 			let mut signed_submissions = Self::signed_submissions();
-			let ejected_a_solution = signed_submissions.len()
-				== T::SignedMaxSubmissions::get().saturated_into::<usize>();
-			let index = Self::insert_submission(&who, &mut signed_submissions, solution, size)
-				.ok_or(Error::<T>::SignedQueueFull)?;
-
-			// collect deposit. Thereafter, the function cannot fail.
-			let deposit = signed_submissions
-				.get(index)
-				.map(|s| s.deposit)
-				.ok_or(Error::<T>::InvalidSubmissionIndex)?;
-			T::Currency::reserve(&who, deposit).map_err(|_| Error::<T>::SignedCannotPayDeposit)?;
-
-			// store the new signed submission.
-			<SignedSubmissions<T>>::put(signed_submissions);
+
+			let (maybe_deposit, ejected_a_solution) =
+				Self::insert_submission(&who, &mut signed_submissions, solution, size);
+			// it's an error if we neither inserted nor removed any submissions
+			ensure!(
+				(None, false) != (maybe_deposit, ejected_a_solution),
+				Error::<T>::SignedQueueFull,
+			);
+			signed_submissions.put();
+
+			if let Some(deposit_amount) = maybe_deposit {
+				// collect deposit. Thereafter, the function cannot fail.
+				T::Currency::reserve(&who, deposit_amount)
+					.map_err(|_| Error::<T>::SignedCannotPayDeposit)?;
+			}
```

Comment:
Note that we haven't really done check-before-write here, because `signed_submissions` is three data structures in a trenchcoat pretending to be one. https://github.com/paritytech/substrate/blob/847b98cf5fa6f3a025b62ae9772eac27a0c531ef/frame/election-provider-multi-phase/src/signed.rs#L188-L191

If the deposit fails, then we end up with an incremented `SignedSubmissionsNextIndex` and an orphan entry in the `SignedSubmissionsMap`. The former is harmless. The latter occupies some chain storage space, but it will be cleared when the election is finalized.

Reply:
Also: https://github.com/paritytech/substrate/pull/8933/commits/5b2a7c5f3fcb939f38249026dd1d8eb42de0f838 already moved the `signed_submissions.put()` to after the deposit.

Reply:
7f4a837ce fixes this.